ui/app/abilities/variable.js

```diff
@@ -41,4 +47,86 @@ export default class Variable extends AbstractAbility {
       return capabilities.includes('create');
     });
   }
+
+  @computed('token.selfTokenPolicies.[]')
+  get allPaths() {
+    return get(this, 'token.selfTokenPolicies')
+      .toArray()
+      .reduce((paths, policy) => {
+        const [variables] = get(policy, 'rulesJSON.Namespaces').mapBy(
+          'SecureVariables'
+        );
+        paths = { ...paths, ...variables };
+        return paths;
+      }, {});
+  }
+
+  _formatMatchingPathRegEx(path, wildCardPlacement = 'end') {
+    const replacer = () => '\\/';
+    if (wildCardPlacement === 'end') {
+      const trimmedPath = path.slice(0, path.length - 1);
+      const pattern = trimmedPath.replace(WILDCARD_PATTERN_REGEX, replacer);
+      return pattern;
+    } else {
+      const trimmedPath = path.slice(1, path.length);
+      const pattern = trimmedPath.replace(WILDCARD_PATTERN_REGEX, replacer);
+      return pattern;
+    }
+  }
+
+  _computeAllMatchingPaths(pathNames, path) {
+    const matches = [];
+
+    for (const pathName of pathNames) {
+      const pathSubString = pathName.match(PATH_PATTERN_REGEX)[1];
+      const sanitizedPath = JSON.parse(pathSubString);
+      const doesEndWithWildcard =
+        sanitizedPath[sanitizedPath.length - 1] === WILDCARD_GLOB;
+      const doesStartWithWildcard = sanitizedPath[0] === WILDCARD_GLOB;
+
+      if (doesEndWithWildcard) {
+        const formattedPath = this._formatMatchingPathRegEx(sanitizedPath);
+        if (path.match(formattedPath)) matches.push(sanitizedPath);
+      } else if (doesStartWithWildcard) {
+        const formattedPath = this._formatMatchingPathRegEx(
+          sanitizedPath,
+          'start'
+        );
+        if (path.match(formattedPath)) matches.push(sanitizedPath);
+      }
+    }
+
+    return matches;
+  }
+
+  _nearestMatchingPath(path) {
+    const formattedPathKey = `Path "${path}"`;
+    const pathNames = Object.keys(this.allPaths);
+
+    if (pathNames.includes(formattedPathKey)) {
+      return path;
+    }
+
+    const allMatchingPaths = this._computeAllMatchingPaths(pathNames, path);
+
+    if (!allMatchingPaths.length) return WILDCARD_GLOB;
+
+    return allMatchingPaths.reduce((matchingPath, currentPath) => {
+      if (matchingPath === '') {
+        matchingPath = currentPath;
+        return matchingPath;
+      }
+      const count = matchingPath.match(WILDCARD_PATTERN_REGEX)?.length || 0;
+      if (currentPath.match(WILDCARD_PATTERN_REGEX)?.length > count) {
+        matchingPath = currentPath;
+      } else if (currentPath.match(WILDCARD_PATTERN_REGEX)?.length === count) {
+        // Chose suffix over prefix
+        if (currentPath.endsWith(WILDCARD_GLOB)) {
+          matchingPath = currentPath;
+        }
+      }
```

Comment:
This algorithm works for the tests that we have but I'm worried that not using the same algorithm we have in [`findAllMatchingWildcards`](https://github.com/hashicorp/nomad/blob/6fb29066d9e24b5921061a22f99303cf07d4523c/acl/acl.go#L421-L445) is going to run into subtle discrepancies between the behavior of the API and the UI. 

The algorithm described there is:
* candidates are sorted
* a path is a candidate if it glob matches
* each candidate has a difference calculated as `len(path) - len(pattern) + (number of globs in pattern)`
* the smallest difference is used, using the first candidate

Trying to implement a separate algorithm seems like a recipe for trying to encode behaviors that are emergent from the original algorithm (like the prefix before suffix thing, which is really as a result of sorting).

Reply:
Alright, I can refactor this. I was trying to follow your and @philrenaud's guidance re:  "Allow by default" behavior. This was the loosest algorithm that I could design to patch that pattern.

But, I could totally use some help turning the 4 parts of the algorithm into specifications that I can test against. Specifically:

- what the sorting pattern is (e.g. ascending by first character of path)
- how to compute `len(pattern`
- number of globs in pattern??? as in `foo/*/bar/*/baz` is a possibility?

Reply:
* The sorting comes out of the [recursive walk](https://github.com/hashicorp/go-immutable-radix/blob/49d1d02c49a783de548d1ba8ae8fde466a20b9e6/node.go#L300-L315) of the [immutable radix tree](https://github.com/hashicorp/go-immutable-radix) data structure, which gets you a lexicographical sort order. See https://play.golang.com/p/Rtzyi8OfnaT for an example (press **Run** on that page).
* `len(pattern)` is the length in bytes of the policy's path
* number of globs in the pattern is yes just a count of `*` characters. More than one is a possibility (although incredibly ill-advised on the part of the operator 😀 )

Reply:
OK, I think writing this algorithm is going to be pretty challenging.

Reply:
Hmmm... so do you think I should just translate this into JavaScript to get the sorting done?

```
func recursiveWalk(n *Node, fn WalkFn) bool {
	// Visit the leaf values if any
	if n.leaf != nil && fn(n.leaf.key, n.leaf.val) {
		return true
	}

	// Recurse on the children
	for _, e := range n.edges {
		if recursiveWalk(e.node, fn) {
			return true
		}
	}
	return false
}
```